docs/design.md

```diff
@@ -0,0 +1,1319 @@
+  just read the value. If the write transaction has not yet been
+  committed, then the reader has two options. If the write conflict
+  is from an SI transaction, the reader can *push that transaction's
+  commit timestamp into the future* (and consequently not have to
+  read it). This is simple to do: the reader just updates the
+  transaction’s commit timestamp to indicate that when/if the
+  transaction does commit, it should use a timestamp *at least* as
+  high. However, if the write conflict is from an SSI transaction,
+  the reader must compare priorities. If it has the higher priority,
+  it pushes the transaction’s commit timestamp, as with SI (that
+  transaction will then notice its timestamp has been pushed, and
+  restart). If it has the lower priority, it retries itself using as
+  a new priority `max(new random priority, conflicting txn’s
+  priority - 1)`.
+
+- **Writer encounters uncommitted write intent with lower priority**:
+  writer aborts the conflicting transaction.
+
+- **Writer encounters uncommitted write intent with higher priority**:
+  the transaction retries, using as a new priority *max(new random
+  priority, conflicting txn’s priority - 1)*; the retry occurs after
+  a short, randomized backoff interval.
+
+- **Writer encounters committed write intent or newer committed value**:
+  The transaction restarts. On restart, the same priority is reused,
+  but the candidate timestamp is moved forward to the encountered
+  value's timestamp.
+
+**Transaction management**
+
+Transactions are managed by the client proxy (or gateway in SQL Azure
+parlance). Unlike in Spanner, writes are not buffered but are sent
+directly to all implicated ranges. This allows the transaction to abort
+quickly if it encounters a write conflict. The client proxy keeps track
+of all written keys in order to cleanup write intents upon transaction
+completion.
+
+If a transaction is completed successfully, all intents are upgraded to
+committed. In the event a transaction is aborted, all written intents
+are deleted. The client proxy doesn’t guarantee it will cleanup intents;
+but dangling intents are upgraded or deleted when encountered by future
+readers and writers and the system does not depend on their timely
+cleanup for correctness.
+
+In the event the client proxy restarts before the pending transaction is
+completed, the dangling transaction would continue to live in the
+transaction table until aborted by another transaction. Transactions
+heartbeat the transaction table every five seconds by default.
+Transactions encountered by readers or writers with dangling intents
+which haven’t been heartbeat within the required interval are aborted.
+
+An exploration of retries with contention and abort times with abandoned
+transaction is
+[here](https://docs.google.com/document/d/1kBCu4sdGAnvLqpT-_2vaTbomNmX3_saayWEGYu1j7mQ/edit?usp=sharing).
+
+**Transaction Table**
+
+```
+enum Isolation {
+
+SNAPSHOT = 1;
+
+SERIALIZABLE = 2;
```

Comment:
I've actually removed it - it's quite out of date. We'll have to figure out how to best do this, but for now I'll link directly to the proto with a hint on what to read.